### ISSUE TITLE
PDF Accessibility (PDF/UA)

### ISSUE THREAD

**Issue by nehasancheti** [bpampuch/pdfmake#942]
Hi, 
This is more of a question. I went through the features it supports. Does it support pdf Accessibilty as mentioned in  "https://www.w3.org/TR/WCAG-TECHS/pdf.html". 
I am mainly looking for : 
1.  Tagging elements that will help create a table of contents. 
2. Screen readers are able to read this document which will contains columns/tables/lists etc. 

If it does, can you guide me on where I can find examples of getting this done?

Thanks,
Neha.

**Comment by GuanQiuQiao** [comment]
You can refer to here  http://pdfmake.org/playground.html

**Comment by achecht** [comment]
I referred to the Playground and specifically the styles and images sections.  There is nothing there about Tagging for screen reader compatibility or setting Alt Text for images.  The PDF spec has handling of both tagging and alt text.  Is it possible that PDFMake might have this capabiity in the future?  If you think it has it now, just give me a simple example of how to add a navigation tag to a text element declaration and how to add alt text for an image element.  Much appreciated!

**Comment by msesun** [comment]
Is this going to be addressed? We also would like an a11y feature

**Comment by wroge** [comment]
Is it already possible to add alternative texts?

**Comment by garosan** [comment]
I came here to ask the same question but it seems that as of today it's not possible to tag a pdf for accessibility compliance with pdfMake :(

**Comment by rcassidy-vanguard** [comment]
My team has this need as well.  I've forked the repository and will be taking a look to see if we can implement this feature.

**Comment by achecht** [comment]
@rcassidy-vanguard: If you do fork, please post the info to this thread.  I'd be happy to take a look/test it for you.

**Comment by tesshut** [comment]
There is a big need for a free HTML to PDF converter to have this feature as none seem to.

I think many people don't know that tags exist or what they do in PDFs. They are essential for assistive technologies (like a screen reader) to read and interact with a PDF. The tag tree can be found in Acrobat Pro > View > Show/Hide > Navigation Panes > Accessibility tags. 

The most common tags map pretty closely to HTML. Heading tags for example are the same, they're <h1>, <h2> etc. It would in theory be a fairly straightforward conversion (though I am not an engineer so take that with a grain of salt).

I'd love to see this feature implemented. If anyone has any luck with a fork please let us know. I remediate PDFs professionally, and if I can be of any help, please reach out.

**Comment by LukaszJaro** [comment]
I have a requirement to generate accessible PDF. The PDFs are about 1 page with checkboxes and text fields, headings.

Is their any support at all for tags? Like heading tags(h1), list tags(ul li), table of contents, etc.

**Comment by martinogarrizzo5** [comment]
Please add accessibility to the library! There are many people nowadays that requires to interact with documents using screen readers. This feature would help so many people!

**Comment by martinogarrizzo5** [comment]
Are there any updates on this proposal?

**Comment by achecht** [comment]
> Are there any updates on this proposal?

Sadly, we haven't seen any formal response to the accessibility/tagging request and no changes to the code (that I know of) to address that accessibility need.  Devs are welcome to correct me if that is not accurate.

**Comment by martinogarrizzo5** [comment]
I'm not sure about the effort that would be required to implement this kind of feature but if this library is based on pdfkit then I think it should be possible to integrate their accessibility system. 
I found this section on their documentation website: [https://pdfkit.org/docs/accessibility.html](https://pdfkit.org/docs/accessibility.html)

**Comment by liborm85** [comment]
Implemented by commit https://github.com/bpampuch/pdfmake/commit/ca6f874dc1c9c66146f7ce56352291b76482e12b and example https://github.com/bpampuch/pdfmake/commit/4e87757a28ba78a6002b0ddf3aa8225acc703579.

**Comment by achecht** [comment]
@liborm85, I'm excited to try this out.  Thanks for your efforts!

**Comment by liborm85** [comment]
Released in version 0.2.15.

**Comment by rivulent** [comment]
@liborm85 or @achecht have you found it possible to specify alt text for images such that the PDF picks them up?  I note that PDFKit indicates this is possible: https://pdfkit.org/docs/accessibility.html

However I have not been able to determine how to accomplish this in PDFMake.

**Comment by axemonkey** [comment]
Hello!

I was excited to see that tagging has been implemented as of version 0.2.15, as I have implemented a PDFMake solution in a work project but have been frustrated (as have others above apparently) by there being no way to make my PDFs accessible with this otherwise brilliant library.

I am now specifying the version, subset, tagged, and other params as detailed in the example. However, I can still see no way to actually implement tagging when creating my PDFs. There appears to be no way to access the PDFKit methods `[markContent()](https://pdfkit.org/docs/accessibility.html#marked_content)` and `[markStructureContent()](https://pdfkit.org/docs/accessibility.html#logical_structure)` nor any equivalent to build up a structured tagged document using PDFMake.

Is this something that is still being worked on? Or is it possible at the moment, and I just haven't figured out how? Thanks!

**Comment by stuartwilkes** [comment]
@axemonkey @liborm85 @achecht We also would like to implement proper tagging of the actual content itself within pdfmake to ensure our PDFs meet accessibility standards. I have started looking at how this is possible within pdfmake utilising the underlying accessibility features of pdfkit. 

For the moment, I've forked pdfmake and have been working on a basic implementation. With relatively few initial changes, I have an implementation that allows for tagging of text content within pdfs, tagging up paragraphs and headers within the PDFs we output from our application.

The changes are mainly implemented in the `renderPages` function within `printer.js` - you can see the initial work here: https://github.com/wellcometrust/pdfmake/blob/815a39e121a17d76c8222d465c43625ce20346d0/src/printer.js#L376

This is a very basic implementation that doesn't correctly nest all of the content right now - but as you can see from the screengrab below - it gets us to having actual tagged content:

<img width="1887" height="1082" alt="Image" src="https://github.com/user-attachments/assets/7b50ab84-cde8-425d-9363-c77f9c2ee1b7" />

There's plenty of limitations and things that aren't working as required as yet, notably:

- content is not correctly nested within the document tag
- lists are tagged as paragraphs, because reliably creating the correct nested structure (as described in the pdfkit docs (https://pdfkit.org/docs/accessibility.html) is a challenge - especially for ordered lists because of the way that pdfmake creates these 
- I've ignored everything other than text content for the moment (I'm tagging images, svgs etc as artifacts so they're ignored in the structure) as for our use case we don't actually need these to be tagged
- Tables are not considered

I'm focussing at the moment on making this work for our use case rather than general pdfs, so this approach might not be suitable for consideration for inclusion in the library. For example - we use HTML to pdfmake to handle HTML conversion for inclusion in our PDFs. In the output from that, line items are given the property `nodeName` with the HTML tag that has been used (i.e. `H1`, `P`, `LI` etc) - I'm using this to tag heading levels, and have amended our actual application to add these properties for other content we add to our PDFs manually, and am using this property to tag headings correctly in the document definition - other users of the library may not be doing this and a more comprehensive approach might consider amending the document definition format or using another property like `headlineLevel` to do this differently.

**Comment by achecht** [comment]
I'm glad to see someone take a fresh look at this.  I'll be interested to watch progress (as posted here or in the forked project).
~A
